torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+
+  // Get inputs for origin expression
+  auto expr_inputs = origin_expr->inputs();
+
+  // Expr* producer_origin =
+  createExprConsumer(origin_expr, producer);
+
+  // Expr* producer_uses =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // Before: This TV -> Next TV
+  // After:  New TV (CB) -> This TV -> Next TV
+  if (this->hasComputeAt()) {
+    TransformReplay::replayPasC(producer, consumer, -1);
+    auto this_ca_pos = this->getThisComputeAtAxis();
+    producer->computeAt(consumer, this_ca_pos);
+  } else {
+    // Before: Prev TV -> This TV
+    // After:  Prev TV -> New TV (CB) -> This TV
+    // Iterate over origin expression inputs for cache_before on outputs
+    for (Val* v : expr_inputs) {
+      if (v->getValType().value() == ValType::TensorView) {
+        TensorView* origin_input = dynamic_cast<TensorView*>(v);
+        if (origin_input->hasComputeAt() &&
+            origin_input->getComputeAtView() == this) {
+          TransformReplay::replayPasC(producer, consumer, -1);
+
+          auto origin_ca_pos = origin_input->getThisComputeAtAxis();
+          auto origin_rel_ca_pos = origin_input->getRelativeComputeAtAxis();
+          origin_input->computeAt(producer, origin_ca_pos);
+          producer->setComputeAt(consumer, origin_rel_ca_pos);
+        }
+      }
+    }
+  }
+
+  return producer;
+}
+
+// Create a TensorView after the original tensor. A common use case is to read
+// original tensor into shared memory or local registers.
+// Analogous to TVM Cache_Read
+TensorView* TensorView::cache_after() {
+  FusionGuard fg(this->fusion());
+
+  // Get all the uses for this Tensorview
+  TORCH_CHECK(
+      !this->fusion()->hasOutput(this),
+      "Error adding cache_after ",
+      this,
+      " we restrict using cache_after on an output.");
+
+  // Create Consumer Domain
+  // Keep Broadcast Axis (Permanent)
+  auto root_domain = getRootDomain();
+  std::vector<IterDomain*> new_root_domain;
+  for (auto root : root_domain) {
+    if (root->isBroadcast()) {
+      new_root_domain.push_back(new IterDomain(
+          root->start(),
+          root->extent(),
+          root->parallel_method(),
+          false,
+          false,
+          true));
+    } else if (!root->isBroadcast() && !root->isReduction()) {
+      new_root_domain.push_back(new IterDomain(
+          root->start(), root->extent(), root->parallel_method()));
+    }
+  }
+
+  // This domain will be the producer, so create the consumer
+  TensorView* consumer = new TensorView(
+      new TensorDomain(new_root_domain), this->getDataType().value());
+
+  // Set domain of producer - No Change
+  TensorView* producer = this;
+
+  // Insert consumer - Cache_After (CA) - after this TV.
+  // Before: This TV -> [Use Op] -> Next TV
+  // After:  This TV -> [Set Op] -> New CA TV -> [Use Op] -> Next TV
+
+  // Expr* consumer_uses =
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
```

Comment:
do we need to make a copy of the uses? how about 
`const std::unordered_set<Expr*>& uses_expr = this->fusion()->unordered_uses(this);`

or
`const auto& uses_expr = this->fusion()->unordered_uses(this);`

... or even better:
`for (auto expr : usion()->unordered_uses(this);) { ... } `

Reply:
`for (auto expr : fusion()->unordered_uses(this);) { ... }` is fine.